Assets/MixedRealityToolkit.SDK/Experimental/Features/Dwell/DwellHandler.cs

```diff
@@ -0,0 +1,210 @@
+namespace Microsoft.MixedReality.Toolkit.Experimental.Dwell
+{
+    [Serializable]
+    public class PointerUnityEvent : UnityEvent<IMixedRealityPointer>
+    {
+    }
+
+    /// <summary>
+    /// Use this component to add a Dwell modality (https://docs.microsoft.com/en-us/windows/mixed-reality/gaze-and-dwell) to the UI target.
+    /// </summary>
+    public class DwellHandler : MonoBehaviour, IMixedRealityFocusHandler
+    {
+        protected enum DwellState
+        {
+            None = 0,
+            FocusGained,
+            DwellIntended,
+            DwellStarted,
+            DwellCompleted,
+            DwellCanceled,
+            Invalid
+        }
+
+        [Header("Dwell Settings")]
+        [SerializeField]
+        protected DwellProfile dwellProfile = null;
+
+        [Header("Dwell Events")]
+        [SerializeField]
+        private PointerUnityEvent DwellIntended = new PointerUnityEvent();
+
+        [SerializeField]
+        private PointerUnityEvent DwellStarted = new PointerUnityEvent();
+
+        [SerializeField]
+        private PointerUnityEvent DwellCompleted = new PointerUnityEvent();
+
+        [SerializeField]
+        private PointerUnityEvent DwellCanceled = new PointerUnityEvent();
+
+        /// <summary>
+        /// Invoke this event if the dwell is interrupted with alternate modality
+        /// </summary>
+        public UnityAction CancelDwellAction;
+
+        protected bool HasFocus { get; private set; }
+
+        /// <summary>
+        /// Cached pointer reference to track focus events maps to teh same pointer that initiated dwell
+        /// </summary>
+        private IMixedRealityPointer pointerRef;
+
+        private DateTime focusEnterTime = DateTime.MaxValue;
+        private DateTime focusExitTime = DateTime.MaxValue;
+
+        protected DwellState currentDwellState = DwellState.None;
+        protected float fillTimer = 0;
+
+        private void Awake()
+        {
+            Debug.Assert(dwellProfile != null, "DwellProfile is null, creating default profile.");
+
+            if (dwellProfile == null)
+            {
+                dwellProfile = ScriptableObject.CreateInstance("DwellProfile") as DwellProfile;
+            }
+
+            CancelDwellAction += OnDwellCancelAction;
+        }
+
+        private void Update()
+        {
+            if (HasFocus && currentDwellState != DwellState.DwellCompleted)
+            {
+                double focusTime = (DateTime.UtcNow - this.focusEnterTime).TotalSeconds;
+
+                if (focusTime >= dwellProfile.DwellIntentDelay && currentDwellState == DwellState.FocusGained)
+                {
+                    DwellIntended.Invoke(pointerRef);
+                    currentDwellState = DwellState.DwellIntended;
+                    fillTimer = 0;
+                }
+                else if ((focusTime - dwellProfile.DwellIntentDelay) >= dwellProfile.DwellStartDelay && currentDwellState == DwellState.DwellIntended)
+                {
```

Comment:
super nit: I'd do the state check prior to the math in this if clause.  And consider refactoring the math slightly to add the delays right of the ">=" sign instead of subtracting some of them
(currentState == DwellIntended && (focusTime >= DwellIntendDelay + DwellStartDelay)